File: studio/src/engine/checkFabric.ts
/**
 * Is the fence actually connected, and is the roof actually a roof? — ported
 * from tools/check_fabric.py.
 *
 * Five checks, each taking a BlockGrid and returning arrays of human-readable
 * finding strings:
 *
 *   fenceFaults(grid)               — fence/pane/bars/wall cells whose connection
 *                                     booleans disagree with the actual grid, plus
 *                                     posts that connect to nothing at all.
 *   roofFaults(grid, minY = 3)      — hanging roof blocks, holes in a roof plane,
 *                                     enclosed floor cells open to the sky.
 *   lineFaults(grid, y = 1)         — diagonal fence steps, duplicate parallel runs.
 *   slabFaults(grid)                — cubes/rails sitting in the empty upper half
 *                                     of a bottom slab's cell (FabricGuard).
 *   cantileverFaults(grid, minY = 2) — roof blocks with no support column within
 *                                     3 cells.
 *
 * Calibrated over the author's 115 building-like NBTs before any of it was
 * believed (see tools/check_fabric.py docstring). The thresholds that gate:
 *
 *   roof blocks hanging ........... HANGING_MAX = 0  (zero across every author file)
 *   gaps in a roof plane .......... HOLES_MAX  = 2  (his worst case, granary_lvl5..7)
 *   slab riders ................... SLAB_RIDERS_MAX = 0
 *   wrong fence props ............. hard fault for generated files only; the author
 *                                   leaves stale states in 29 of his hand-built files.
 *
 * Counted for information, never failed:
 *   rails connecting to nothing — his own idiom (36 in merchant_shop_lvl6);
 *   enclosed cells open to sky  — courtyards, open work bays.
 */

import { type BlockGrid, type BlockInfo } from './appearance';
import { FabricGuard, type Coord } from './fabric';

// ── constants (from pasture.py, measured off the author's 121 files) ──────

/** Blocks a fence connects to. */
const STURDY = new Set([
  'oak_planks', 'oak_log', 'cobblestone', 'mossy_cobblestone', 'stone',
  'stripped_oak_log', 'hay_block', 'white_terracotta', 'crafting_table',
  'andesite', 'stone_bricks',
]);

/** Direction → (dx, dz). */
const VEC: Record<string, [number, number]> = {
  north: [0, -1], south: [0, 1], west: [-1, 0], east: [1, 0],
};

/** Facing → axis label (only equality/inequality is ever tested). */
const AXIS_OF: Record<string, string> = {
  north: 'ns', south: 'ns', west: 'ew', east: 'ew',
};

const RAIL_SUFFIX = ['_fence', '_pane', '_bars'];
const DIRECTIONS = ['north', 'south', 'east', 'west'] as const;
const NEIGH4: ReadonlyArray<[number, number]> = [[1, 0], [-1, 0], [0, 1], [0, -1]];
const DIAGONALS: ReadonlyArray<[number, number]> = [[1, 1], [1, -1], [-1, 1], [-1, -1]];
const PARALLEL: ReadonlyArray<[number, number]> = [[2, 0], [0, 2]];

/** Too thin or too small to bridge a corner or fill a dead cell. */
const SIDE_ATTACHED_SOFT = new Set([
  'short_grass', 'grass', 'oak_sapling', 'flower_pot', 'torch',
  'wall_torch', 'lantern', 'chain', 'dandelion', 'poppy',
]);

// Thresholds (measured over the author's corpus; only three of the five gate).
const HANGING_MAX = 0;
const HOLES_MAX = 2;
const SLAB_RIDERS_MAX = 0;

function compareCoord(a: Coord, b: Coord): number {
  return a[0] - b[0] || a[1] - b[1] || a[2] - b[2];
}

function endsWithAny(s: string, suffixes: readonly string[]): boolean {
  return suffixes.some(suffix => s.endsWith(suffix));
}

function isRoofMaterialId(id: string): boolean {
  return id.endsWith('_slab') || id.endsWith('_stairs');
}

/** A roof neighbour for the hole-ring check: slabs, stairs, planks, hay. */
function isRoofBlock(b: BlockInfo | null): boolean {
  return b !== null && (isRoofMaterialId(b.id) || b.id === 'oak_planks' || b.id === 'hay_block');
}

/** Sorted view of grid.blocks() by (x, y, z) — deterministic output order. */
function sortedBlocks(grid: BlockGrid): Array<{ pos: Coord; block: BlockInfo }> {
  return [...grid.blocks()].sort((a, b) => compareCoord(a.pos, b.pos));
}

// ── fences ────────────────────────────────────────────────────────────────

/**
 * Would a rail at `p` connect toward `direction` in the actual game?
 *
 * Same rule `pasture.reconnect` derives from, measured off the author's corpus —
 * including the part that says a fence and a `*_wall` block do NOT connect.
 */
function links(grid: BlockGrid, p: Coord, direction: string, family: 'rail' | 'wall'): boolean {
  const [dx, dz] = VEC[direction];
  const nb = grid.get(p[0] + dx, p[1], p[2] + dz);
  if (nb === null) return false;
  const n = nb.id;
  if (STURDY.has(n)) return true;
  if (n.endsWith('_fence_gate')) {
    return AXIS_OF[nb.props['facing'] ?? 'north'] !== AXIS_OF[direction];
  }
  if (family === 'rail') {
    return endsWithAny(n, RAIL_SUFFIX);
  }
  return n.endsWith('_wall');
}

/**
 * Cells whose connection props are wrong, and posts connecting to nothing.
 * A rail-to-nothing post is the author's own idiom — reported, not failed.
 */
export function fenceFaults(grid: BlockGrid): { wrong: string[]; stumps: string[] } {
  const wrong: string[] = [];
  const stumps: string[] = [];
  for (const { pos: p, block: b } of sortedBlocks(grid)) {
    const n = b.id;
    if (!(endsWithAny(n, RAIL_SUFFIX) || n.endsWith('_wall'))) continue;
    const family: 'rail' | 'wall' = n.endsWith('_wall') ? 'wall' : 'rail';
    const want: Record<string, boolean> = {};
    const have: Record<string, boolean> = {};
    for (const d of DIRECTIONS) {
      want[d] = links(grid, p, d, family);
      have[d] = n.endsWith('_wall')
        ? (b.props[d] ?? 'none') !== 'none'
        : (b.props[d] ?? 'false') === 'true';
    }
    if (DIRECTIONS.some(d => have[d] !== want[d])) {
      const bad = DIRECTIONS.filter(d => have[d] !== want[d]);
      const hasList = DIRECTIONS.filter(d => have[d]);
      const wantList = DIRECTIONS.filter(d => want[d]);
      wrong.push(
        `${n}@(${p.join(', ')}) ${bad.join(',')}: has [${hasList.join(', ')}] wants [${wantList.join(', ')}]`,
      );
    }
    if (!DIRECTIONS.some(d => want[d])) {
      stumps.push(`${n}@(${p.join(', ')})`);
    }
  }
  return { wrong, stumps };
}

// ── roofs ─────────────────────────────────────────────────────────────────

/** Hanging blocks, holes in a plane, and enclosed floor cells open to the sky. */
export function roofFaults(grid: BlockGrid, minY = 3): {
  hanging: string[];
  holed: string[];
  sky: string[];
} {
  const hanging: string[] = [];
  const holed: string[] = [];
  const sky: string[] = [];

  const solid = new Set<string>();
  for (const { pos } of grid.blocks()) solid.add(`${pos[0]},${pos[1]},${pos[2]}`);
  const inSolid = (x: number, y: number, z: number): boolean => solid.has(`${x},${y},${z}`);

  for (const { pos: p, block: b } of sortedBlocks(grid)) {
    if (p[1] < minY || !isRoofMaterialId(b.id)) continue;
    if (inSolid(p[0], p[1] - 1, p[2])) continue;
    if (NEIGH4.some(([dx, dz]) => inSolid(p[0] + dx, p[1], p[2] + dz))) continue;
    const stepped = NEIGH4.some(
      ([dx, dz]) => inSolid(p[0] + dx, p[1] - 1, p[2] + dz) || inSolid(p[0] + dx, p[1] + 1, p[2] + dz),
    );
    if (stepped) continue;
    hanging.push(`${b.id}@(${p.join(', ')})`);
  }

  // A hole: empty cell with roof on three or more sides at its own height.
  const [sx, sy, sz] = grid.size;
  for (let y = minY; y < sy; y++) {
    for (let x = 1; x < sx - 1; x++) {
      for (let z = 1; z < sz - 1; z++) {
        if (inSolid(x, y, z)) continue;
        let ring = 0;
        for (const [dx, dz] of NEIGH4) {
          if (isRoofBlock(grid.get(x + dx, y, z + dz))) ring++;
        }
        if (ring >= 3 && !grid.occupied(x, y + 1, z)) {
          holed.push(`gap@(${x}, ${y}, ${z}) roof on ${ring} sides`);
        }
      }
    }
  }

  // Enclosed floor cells with nothing above them. Normal in the corpus —
  // courtyards, open work bays — so counted rather than failed.
  for (let x = 1; x < sx - 1; x++) {
    for (let z = 1; z < sz - 1; z++) {
      if (grid.occupied(x, 1, z)) continue;
      const walled = NEIGH4.every(([dx, dz]) => grid.occupied(x + dx, 1, z + dz));
      let roofed = false;
      for (let y = 2; y < sy; y++) {
        if (grid.occupied(x, y, z)) {
          roofed = true;
          break;
        }
      }
      if (walled && !roofed) {
        sky.push(`enclosed cell open to the sky@(${x}, 1, ${z})`);
      }
    }
  }
  return { hanging, holed, sky };
}

// ── the boundary line ─────────────────────────────────────────────────────

/**
 * Two ways a fence line reads as broken even when it holds animals:
 *   - diagonal step — two rails meeting only at a corner;
 *   - duplicate run — two parallel lines one cell apart with a dead cell between.
 *
 * Anything solid in the corner cell bridges the step visually (a slab, a
 * trapdoor, a bed). Requiring a *barrier* there reported 14 false positives
 * inside the author's own house; the question is whether the run looks broken,
 * not whether a cow could squeeze through.
 */
export function lineFaults(grid: BlockGrid, y = 1): { diagonal: string[]; duplicate: string[] } {
  const diagonal: string[] = [];
  const duplicate: string[] = [];
  const rails = new Set<string>();
  const barrier = new Set<string>();
  for (const { pos: p, block: b } of grid.blocks()) {
    if (p[1] !== y) continue;
    const id = b.id;
    if (id.endsWith('_fence') || id.endsWith('_fence_gate') || id.endsWith('_wall')) {
      rails.add(`${p[0]},${p[2]}`);
    }
    if (!SIDE_ATTACHED_SOFT.has(id)) {
      barrier.add(`${p[0]},${p[2]}`);
    }
  }
  const railCoords = [...rails]
    .map(s => s.split(',').map(Number) as [number, number])
    .sort((a, b) => a[0] - b[0] || a[1] - b[1]);

  for (const [x, z] of railCoords) {
    for (const [dx, dz] of DIAGONALS) {
      const qx = x + dx;
      const qz = z + dz;
      if (!rails.has(`${qx},${qz}`)) continue;
      // A shared orthogonal neighbour makes the two part of one run.
      if (barrier.has(`${x + dx},${z}`) || barrier.has(`${x},${z + dz}`)) continue;
      diagonal.push(`(${x}, ${z}) meets (${qx}, ${qz}) only diagonally`);
    }
  }

  const seen = new Set<string>();
  for (const [x, z] of railCoords) {
    for (const [dx, dz] of PARALLEL) {
      const qx = x + dx;
      const qz = z + dz;
      const mx = x + Math.floor(dx / 2);
      const mz = z + Math.floor(dz / 2);
      const midKey = `${mx},${mz}`;
      const seenKey = `${midKey}|${qx},${qz}`;
      if (rails.has(`${qx},${qz}`) && !barrier.has(midKey) && !seen.has(seenKey)) {
        seen.add(seenKey);
        duplicate.push(
          `(${x}, ${z}) and (${qx}, ${qz}) run parallel with (${mx}, ${mz}) dead between`,
        );
      }
    }
  }
  return { diagonal, duplicate };
}

// ── half blocks ───────────────────────────────────────────────────────────

/**
 * Cubes and rails sitting in the empty upper half of a bottom slab's cell.
 *
 * Delegated to `FabricGuard`, which is the same code that refuses the write in
 * the first place — one authority, so the writer and the checker cannot drift.
 */
export function slabFaults(grid: BlockGrid): string[] {
  const guard = new FabricGuard();
  return guard.inspectAll(grid).map(f => `${f.kind}: ${f.detail} at (${f.pos.join(', ')})`);
}

// ── cantilevers ───────────────────────────────────────────────────────────

/**
 * Roof blocks too far from anything holding them up.
 *
 * Measured over the author's 121 files: a roof block with no column under it is
 * 1 cell from a supported column 2076 times, 2 cells 355 times, 3 cells ten
 * times and 4 cells once (`oven.nbt`). A deep eave is his idiom; a shelf
 * floating four cells out is not. Here we report blocks with NO supported
 * column within 3 cells.
 */
export function cantileverFaults(grid: BlockGrid, minY = 2): string[] {
  const out: string[] = [];
  const solid = new Set<string>();
  for (const { pos } of grid.blocks()) solid.add(`${pos[0]},${pos[1]},${pos[2]}`);
  const inSolid = (x: number, y: number, z: number): boolean => solid.has(`${x},${y},${z}`);
  const supported = (x: number, z: number, y: number): boolean => {
    for (let yy = 1; yy < y; yy++) {
      if (inSolid(x, yy, z)) return true;
    }
    return false;
  };

  for (const { pos: p, block: b } of sortedBlocks(grid)) {
    const [x, y, z] = p;
    if (y < minY || !isRoofMaterialId(b.id)) continue;
    if (supported(x, z, y)) continue;
    let reach: number | null = null;
    search: for (let r = 1; r <= 3; r++) {
      for (let dx = -r; dx <= r; dx++) {
        const dz = r - Math.abs(dx);
        const cands: ReadonlyArray<[number, number]> = [[x + dx, z + dz], [x + dx, z - dz]];
        for (const [cx, cz] of cands) {
          if (inSolid(cx, y, cz) && supported(cx, cz, y)) {
            reach = r;
            break search;
          }
        }
      }
    }
    if (reach === null) {
      out.push(`${b.id}@(${p.join(', ')}) has no support within 3 cells`);
    }
  }
  return out;
}

// ── combined report ───────────────────────────────────────────────────────

export type FabricFaultBucket = { label: string; items: string[] };

export type FabricReport = {
  fence: { wrong: string[]; stumps: string[] };
  roof: { hanging: string[]; holed: string[]; sky: string[] };
  line: { diagonal: string[]; duplicate: string[] };
  slab: string[];
  cantilever: string[];
  /** Buckets that exceed their measured threshold — the actual failures. */
  faults: FabricFaultBucket[];
};

/**
 * Run all five fabric checks and return a structured report. `faults` holds
 * only the buckets whose count crosses its measured threshold; the raw arrays
 * are kept on the report for the UI and for information-only counts.
 */
export function checkFabric(grid: BlockGrid): FabricReport {
  const fence = fenceFaults(grid);
  const roof = roofFaults(grid);
  const line = lineFaults(grid);
  const slab = slabFaults(grid);
  const cantilever = cantileverFaults(grid);

  const faults: FabricFaultBucket[] = [];
  if (cantilever.length > 0) faults.push({ label: 'cantilever', items: cantilever });
  if (slab.length > SLAB_RIDERS_MAX) faults.push({ label: 'slab-rider', items: slab });
  // Wrong fence props are a hard fault for generated files only; the author
  // leaves stale states in 29 of his own hand-built files.
  if (fence.wrong.length > 0) faults.push({ label: 'fence-props', items: fence.wrong });
  if (roof.hanging.length > HANGING_MAX) faults.push({ label: 'roof-hanging', items: roof.hanging });
  if (roof.holed.length > HOLES_MAX) faults.push({ label: 'roof-holes', items: roof.holed });

  return { fence, roof, line, slab, cantilever, faults };
}

File: tools/check_fabric.py
"""Is the fence actually connected, and is the roof actually a roof?

Run from tools/:

    python check_fabric.py                    # the livestock set
    python check_fabric.py --calibrate        # the author's own builds first
    python check_fabric.py path/to/file.nbt

Two questions the style gate cannot answer and the escape model does not care
about, both raised on review:

**The fence must always connect.** A fence carries `north/south/east/west`
booleans. Placed with the wrong ones it renders as a line of loose stumps with
gaps of air between the rails — the boundary still holds animals, so no
functional check complains, and it looks broken. `pasture.reconnect` derives
these from the finished grid; this re-derives them from the *shipped file* and
reports any cell that disagrees, plus any post that connects to nothing at all.

**The roof must be whole**: no block hanging with nothing under it, and no gap
left in the middle of a plane.

**Calibrated over the author's 115 building-like NBTs before any of it was
believed**, which changed three of the five tests:

| metric | author max | p95 | median | files with any |
|---|---|---|---|---|
| roof blocks hanging | **0** | 0 | 0 | **0 / 115** |
| gaps in a roof plane | 2 | 1 | 0 | 22 / 115 |
| fence props disagreeing | 20 | 4 | 0 | 29 / 115 |
| rails connecting to nothing | 36 | 19 | 5 | 99 / 115 |
| enclosed cells open to the sky | 7 | 5 | 0 | 25 / 115 |

So:

* **hanging is a hard fault** — zero across every file the author ever built,
  which makes it the sharpest test in this repo.
* **gaps are a fault above 2**, his worst case (`granary_lvl5..7`).
* **wrong fence props are a hard fault for generated files only.** The author
  has them in 29 of his own — he built by hand and left stale states behind —
  but we place every block deliberately, so ours must be exact.
* **a rail connecting to nothing is NOT a fault.** It is his own idiom: a
  free-standing post, a leg under a projecting element, 36 of them in
  `merchant_shop_lvl6`. Reported for information only.
* An enclosed cell open to the sky is likewise normal — courtyards and open work
  bays — so it is counted, not failed.

A first version of this file also failed a *lean-to course that meets nothing
inward*, and reported 35 of them in `house_lvl6`. That was the checker being
wrong about how a pitch is built, not the author leaving a roof unfinished; it is
gone. Measure first, then assert.
"""

from __future__ import annotations

import argparse
import sys
from pathlib import Path
from typing import Dict, List, Optional, Sequence, Tuple

sys.path.insert(0, str(Path(__file__).resolve().parent))

from structures.fabric import Canvas
from structures.nbtio import BlockState, Coord, Voxels, load
from structures.pasture import FULL_BLOCKS, NEIGH4, STURDY, VEC, AXIS_OF

LIVESTOCK = Path("../common/src/main/resources/data/burg/structure/livestock")
CORPUS = Path("../common/src/main/resources/data/burg/structure")

ROOF_MATERIAL = ("_slab", "_stairs")
# Too thin or too small to bridge a corner or to fill a dead cell.
SIDE_ATTACHED_SOFT = {"short_grass", "grass", "oak_sapling",
                      "flower_pot", "torch", "wall_torch", "lantern",
                      "chain", "dandelion", "poppy"}
RAIL_SUFFIX = ("_fence", "_pane", "_bars")


# ── fences ──────────────────────────────────────────────────────────

def _links(vox: Voxels, p: Coord, direction: str, family: str) -> bool:
    """Would a rail at `p` connect toward `direction` in the actual game?

    The same rule `pasture.reconnect` derives from, and it is measured off the
    author's corpus rather than guessed — including the part that says a fence
    and a `*_wall` block do not connect to each other.
    """
    dx, dz = VEC[direction]
    nb = vox.get((p[0] + dx, p[1], p[2] + dz))
    if nb is None:
        return False
    n = nb.short
    if n in STURDY:
        return True
    if n.endswith("_fence_gate"):
        return AXIS_OF[nb.get("facing", "north")] != AXIS_OF[direction]
    if family == "rail":
        return n.endswith(("_fence", "_pane", "_bars"))
    return n.endswith("_wall")


def fence_faults(vox: Voxels) -> Tuple[List[str], List[str]]:
    """(cells whose connection props are wrong, posts connecting to nothing)."""
    wrong: List[str] = []
    stumps: List[str] = []
    for p, b in sorted(vox.solid_items()):
        n = b.short
        if not n.endswith(RAIL_SUFFIX + ("_wall",)):
            continue
        family = "wall" if n.endswith("_wall") else "rail"
        want = {d: _links(vox, p, d, family)
                for d in ("north", "south", "east", "west")}
        if n.endswith("_wall"):
            have = {d: b.get(d, "none") != "none" for d in want}
        else:
            have = {d: b.get(d, "false") == "true" for d in want}
        if have != want:
            bad = [d for d in want if have[d] != want[d]]
            wrong.append(f"{n}@{p} {','.join(bad)}: has "
                         f"{[d for d in want if have[d]]} wants "
                         f"{[d for d in want if want[d]]}")
        if not any(want.values()):
            stumps.append(f"{n}@{p}")
    return wrong, stumps


# ── roofs ───────────────────────────────────────────────────────────

def _is_roof(b: Optional[BlockState]) -> bool:
    return b is not None and (b.short.endswith(ROOF_MATERIAL)
                              or b.short in ("oak_planks", "hay_block"))


def roof_faults(vox: Voxels, min_y: int = 3) -> Dict[str, List[str]]:
    """Hanging blocks, holes in a plane, and lean-to courses short of the wall."""
    out: Dict[str, List[str]] = {"hanging": [], "holed": [], "sky": []}
    solid = {p for p, b in vox.solid_items()}

    for p, b in sorted(vox.solid_items()):
        if p[1] < min_y or not b.short.endswith(ROOF_MATERIAL):
            continue
        below = (p[0], p[1] - 1, p[2])
        lateral = [(p[0] + dx, p[1], p[2] + dz) for dx, dz in NEIGH4]
        stepped = [(p[0] + dx, p[1] + dy, p[2] + dz)
                   for dx, dz in NEIGH4 for dy in (-1, 1)]
        if below in solid:
            continue
        if any(q in solid for q in lateral):
            continue
        if any(q in solid for q in stepped):
            continue
        out["hanging"].append(f"{b.short}@{p}")

    # A hole: empty cell with roof on three or more sides at its own height.
    sx, sy, sz = vox.size
    for y in range(min_y, sy):
        for x in range(1, sx - 1):
            for z in range(1, sz - 1):
                if (x, y, z) in solid:
                    continue
                ring = sum(1 for dx, dz in NEIGH4
                           if _is_roof(vox.get((x + dx, y, z + dz))))
                if ring >= 3 and not vox.occupied((x, y + 1, z)):
                    out["holed"].append(f"gap@{(x, y, z)} roof on {ring} sides")

    # Enclosed floor cells with nothing above them. Normal in the corpus —
    # courtyards, open work bays — so counted rather than failed.
    for x in range(1, sx - 1):
        for z in range(1, sz - 1):
            if vox.occupied((x, 1, z)):
                continue
            walled = all(vox.occupied((x + dx, 1, z + dz)) for dx, dz in NEIGH4)
            roofed = any(vox.occupied((x, y, z)) for y in range(2, sy))
            if walled and not roofed:
                out["sky"].append(f"enclosed cell open to the sky@{(x, 1, z)}")
    return out


# ── the boundary line ───────────────────────────────────────────────

def line_faults(vox: Voxels, y: int = 1) -> Dict[str, List[str]]:
    """Two ways a fence line reads as broken even when it holds animals.

    * **diagonal step** — two rails meeting only at a corner. Nothing can walk
      through it, so the escape model is happy, and it looks like a gap in the
      run because a fence connects to nothing diagonally.
    * **duplicate run** — two parallel lines of barrier one cell apart with a
      dead cell between them. Usually means one of the two was built against a
      wall that is not where the code thought it was.
    """
    out: Dict[str, List[str]] = {"diagonal": [], "duplicate": []}
    rails = {(p[0], p[2]) for p, b in vox.solid_items()
             if p[1] == y and b.short.endswith(("_fence", "_fence_gate", "_wall"))}
    # Anything solid in the corner cell bridges the step visually — a slab, a
    # trapdoor, a bed. Requiring a *barrier* there reported 14 false positives,
    # every one inside the author's own house where his fences meet across a slab
    # or a trapdoor. The question this metric asks is whether the run looks
    # broken, not whether a cow could squeeze through.
    barrier = {(p[0], p[2]) for p, b in vox.solid_items()
               if p[1] == y and b.short not in SIDE_ATTACHED_SOFT}
    for (x, z) in sorted(rails):
        for dx, dz in ((1, 1), (1, -1), (-1, 1), (-1, -1)):
            q = (x + dx, z + dz)
            if q not in rails:
                continue
            # A shared orthogonal neighbour makes the two part of one run.
            if (x + dx, z) in barrier or (x, z + dz) in barrier:
                continue
            out["diagonal"].append(f"{(x, z)} meets {q} only diagonally")
    seen = set()
    for (x, z) in sorted(rails):
        for dx, dz in ((2, 0), (0, 2)):
            q = (x + dx, z + dz)
            mid = (x + dx // 2, z + dz // 2)
            if q in rails and mid not in barrier and (mid, q) not in seen:
                seen.add((mid, q))
                out["duplicate"].append(
                    f"{(x, z)} and {q} run parallel with {mid} dead between")
    return out


# ── half blocks ─────────────────────────────────────────────────────

# A bottom slab fills the **lower** half of its cell, so a cube in the cell above
# it hangs half a block clear of what is supposedly holding it up. Measured over
# the author's 121 files: 6 occurrences in total, max 2 per file, and every one is
# an `oak_trapdoor` — which attaches to a side and reads fine. So cubes resting on
# a bottom slab are a fault, and side-attached blocks are not.
#
# The reverse — a top slab with nothing beneath it — is NOT a fault: he has 2846
# of them, up to 90 in one file, because a top slab is a step, a table top and a
# railing surface in its own right.
SLAB_RIDERS_MAX = 0
SIDE_ATTACHED = {"oak_trapdoor", "spruce_trapdoor", "iron_trapdoor", "torch",
                 "wall_torch", "lantern", "ladder", "oak_wall_sign", "chain",
                 "tripwire_hook", "white_wall_banner", "red_wall_banner",
                 "brown_wall_banner", "short_grass", "grass", "flower_pot"}


def slab_faults(vox: Voxels) -> List[str]:
    """Cubes and rails sitting in the empty upper half of a bottom slab's cell.

    Delegated to `fabric.Canvas`, which is the same code that refuses the write in
    the first place. It used to be a second implementation here, with its own list of
    side-attached ids, and two copies of a rule are two rules: the writer once
    rejected a fence over a fence that this one allowed. One authority, checked by
    `calibrate_fabric.py --selftest`.
    """
    return [f"{f.kind}: {f.detail} at {f.pos}"
            for f in Canvas(vox).inspect_all()]


def cantilever_faults(vox: Voxels, min_y: int = 2) -> List[str]:
    """Roof blocks too far from anything holding them up.

    Measured over the author's 121 files: a roof block with no column under it is
    1 cell from a supported column 2076 times, 2 cells 355 times, 3 cells ten
    times and 4 cells once (`oven.nbt`). So a deep eave is his idiom and a shelf
    floating four cells out is not. Mine had two blocks with **no** supported
    column within four cells — genuinely hanging in the air.
    """
    solid = {p for p, b in vox.solid_items()}

    def supported(x: int, z: int, y: int) -> bool:
        return any((x, yy, z) in solid for yy in range(1, y))

    out: List[str] = []
    for p, b in sorted(vox.solid_items()):
        x, y, z = p
        if y < min_y or not b.short.endswith(ROOF_MATERIAL):
            continue
        if supported(x, z, y):
            continue
        reach = None
        for r in range(1, 4):
            for dx in range(-r, r + 1):
                dz = r - abs(dx)
                for cand in ((x + dx, y, z + dz), (x + dx, y, z - dz)):
                    if cand in solid and supported(cand[0], cand[2], y):
                        reach = r
                        break
                if reach:
                    break
            if reach:
                break
        if reach is None:
            out.append(f"{b.short}@{p} has no support within 3 cells")
    return out


# ── reporting ───────────────────────────────────────────────────────

# Thresholds, measured above. Only three of the five gate anything.
HANGING_MAX = 0
HOLES_MAX = 2


def report(paths: Sequence[Path], strict_props: bool = True) -> int:
    bad = 0
    for p in paths:
        vox = load(p)
        vox.name = p.stem
        wrong, stumps = fence_faults(vox)
        roof = roof_faults(vox)
        riders = slab_faults(vox)
        floaters = cantilever_faults(vox)
        faults = []
        if floaters:
            faults.append(("cantilever", floaters))
        if len(riders) > SLAB_RIDERS_MAX:
            faults.append(("slab-rider", riders))
        if strict_props and wrong:
            faults.append(("fence-props", wrong))
        if len(roof["hanging"]) > HANGING_MAX:
            faults.append(("roof-hanging", roof["hanging"]))
        if len(roof["holed"]) > HOLES_MAX:
            faults.append(("roof-holes", roof["holed"]))
        print(f"  {'OK   ' if not faults else 'FAULT'} {p.name:26s} "
              f"props={len(wrong)} hanging={len(roof['hanging'])} "
              f"holes={len(roof['holed'])} slab_riders={len(riders)} cantilever={len(floaters)}   "
              f"(info: rails-to-nothing={len(stumps)}, "
              f"open-to-sky={len(roof['sky'])})")
        for label, items in faults:
            for it in items[:4]:
                print(f"          {label}: {it}")
        if faults:
            bad += 1
    print(f"\n{len(paths) - bad}/{len(paths)} clean")
    return bad


def main(argv: Sequence[str]) -> int:
    ap = argparse.ArgumentParser(description=__doc__,
                                 formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("files", nargs="*")
    ap.add_argument("--calibrate", action="store_true",
                    help="run the author's own builds first")
    a = ap.parse_args(argv)

    if a.calibrate:
        print("=== the author's own builds (the checker must be quiet here) ===")
        ref = [CORPUS / f"plains/houses/{n}.nbt" for n in
               ("house", "house_lvl3", "house_lvl6", "house_2_lvl4")]
        ref += [CORPUS / f"plains/jobs/{n}.nbt" for n in
                ("pig_farm_lvl3", "sheep_field", "cow_field", "carpenter_lvl4")]
        report([p for p in ref if p.exists()], strict_props=False)
        print()

    paths = [Path(f) for f in a.files] or sorted(LIVESTOCK.rglob("*.nbt"))
    print("=== the livestock set ===")
    return 1 if report(paths) else 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
